### PR TITLE
Refactor thermal scattering data

### PR DESCRIPTION
**Pull request by paulromano**
In this branch, I've refactored and improved how we handle thermal scattering data to be more in line with the rest of the nuclear data hierarchy (originally discussed in #1233). The main goal was to make the secondary angle/energy distributions for thermal scattering conform to the some class structure as the distributions used for other neutron reactions (subclasses of `AngleEnergy`). To outline the changes:

- The `ThermalScattering` Python class now has attributes `elastic` and `inelastic` that are instances of `ThermalScatteringReaction`.
- The `ThermalScatteringReaction` class has `xs` and `distribution` attributes, the former of which is a 1D function and the latter of which is an instance of `AngleEnergy`.
- There are a bunch of new subclasses of `AngleEnergy`:
  - `CoherentElasticAE` -- coherent elastic scattering (can be generated from ENDF or ACE data)
  - `IncoherentElasticAE` -- incoherent elastic scattering (can be generated from ENDF or ACE data)
  - `IncoherentElasticAEDiscrete` -- same as above but with discrete representation of angles (specifically for ACE data)
  - `IncoherentInelasticAE` -- incoherent inelastic scattering (newer continuous E representation in ACE)
  - `IncoherentInelasticAEDiscrete` -- incoherent inelastic scattering (older discrete E/mu representation, includes skewed distribution)
- The `ThermalScattering` class has a `from_endf` classmethod now. It can't be used for incoherent inelastic data since that requires further processing, but it is possible to pull coherent/incoherent elastic data directly from the ENDF file now. To reflect this, the existing `from_njoy` has a flag to use data directly from ENDF for incoherent elastic, replacing the ACE data.
- Similar changes to classes on the C++ have been made

As usual, during the process of refactor the classes, I came upon a bug that is now fixed. In particular, I found that there was assumption in the C++ code that for coherent elastic scattering, if you were above the highest Bragg edge but below the maximum energy for the S(a,b) table overall, the elastic xs was zero (when in reality it is not).

**Important note**: Because of the fundamental changes in how we are storing thermal scattering data, it was necessary to increment the version of the HDF5 data. This means that after this PR is merged, you will need to download new data in order for OpenMC to work.

A few unrelated changes also:

- bug fix for #1221
- Store the threshold index for neutrons as 0-based (this was one remaining holdout from the 1-based index days). As long as we're incrementing the HDF5 version, we might as well fix this too.

### CODE REVIEW

File: include/openmc/secondary_thermal.h [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+//==============================================================================
+//! Incoherent elastic scattering angle-energy distribution (discrete)
+//==============================================================================
+
+class IncoherentElasticAEDiscrete : public AngleEnergy {
+public:
+  explicit IncoherentElasticAEDiscrete(hid_t group, const std::vector<double>& energy);
+
+  void sample(double E_in, double& E_out, double& mu) const override;
```
[the remark this answers is not in the document]
Comment (paulromano) [reply include/openmc/secondary_thermal.h]: Yes, I'll put appropriate doxygen comments here. Thanks for the reminder!

**Comment by paulromano** [comment]
@nelsonag Yes, I've tried to choose data to cover all the different cases. If you look at the [coverage](https://coveralls.io/builds/24229071) report, you can see that on the Python side our numbers are very high. On the C++ side, they are pretty low, but I'm planning on addressing that once we update our test data to use something other than the NNDC data we've been relying on.

File: docs/source/pythonapi/data.rst [hunk 2]
```diff
@@ -13,18 +13,38 @@ Core Classes
     openmc.data.IncidentNeutron
```
[the remark this answers is not in the document]
Comment (paulromano) [reply docs/source/pythonapi/data.rst]: Will do

File: src/secondary_thermal.cpp [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+  int i;
+  double f;
+  get_energy_index(energy_, E_in, i, f);
+
+  // Sample between ith and [i+1]th bin
+  int l = f > prn() ? i + 1 : i;
+
+  // Determine endpoints on grid i
+  auto n = distribution_[i].e_out.size();
+  double E_i_1 = distribution_[i].e_out(0);
```
[the remark this answers is not in the document]
Comment (paulromano) [reply src/secondary_thermal.cpp]: Yup, I'll change that

**Comment by nelsonag** [comment]
Looks good, thanks @paulromano!

**Pull request merged**